component/espejote-templates/egress-gateway-self-service.jsonnet

```diff
@@ -0,0 +1,109 @@
+local esp = import 'espejote.libsonnet';
+
+local config = import 'egress-gateway/config.json';
+local egw = import 'egress-gateway/egress-gateway.libsonnet';
+local ipcalc = import 'egress-gateway/ipcalc.libsonnet';
+
+// setAnnotations is a helper that generates a minimal partial manifest to
+// set/update annotations with server-side apply.
+local setAnnotations(obj, annotations) = {
+  apiVersion: obj.apiVersion,
+  kind: obj.kind,
+  metadata: {
+    name: obj.metadata.name,
+    annotations: annotations,
+  },
+};
+
+// Collect list of namespaces for which we currently manage egress policies
+// based on the `egress_policies` context (which only contains
+// `IsovalentEgressGatewayPolicy` resources which have the `egw.espejoteLabel`
+// set).
+local managed_policies_namespaces = [
+  p.metadata.name
+  for p in esp.context().egress_policies
+];
+
+local reconcileNamespace(namespace) =
+  assert
+    namespace != null && namespace.kind == 'Namespace'
+    : 'reconcileNamespace() expects to be called with a Namespace resource';
+
+  local ns_meta = namespace.metadata;
+  local egress_ip = std.get(
+    std.get(ns_meta, 'annotations', {}),
+    'cilium.syn.tools/egress-ip'
+  );
+
+  if egress_ip != null then (
+    local res = egw.find_egress_range(config.egress_ranges, egress_ip);
+    if res.range != null then
+      // when we have a range, generate a IsovalentEgressGatewayPolicy (with
+      // ownerReference pointing to the namespace, and labeled as managed by
+      // us) and update the namespace with an informational message.
+      local range = res.range;
+      [
+        egw.NamespaceEgressPolicy(
+          range.if_prefix,
+          range.egress_range,
+          std.objectValues(range.shadow_ranges),
+          range.node_selector,
+          egress_ip,
+          ns_meta.name,
+          egw.IsovalentEgressGatewayPolicy
+        ) {
+          metadata+: {
+            labels+: egw.espejoteLabel,
+            ownerReferences: [ {
```

Comment:
`espejote.libsonnet` should have a helper for that.

Reply:
it feels like `setAnnotations()` (or something similar that's a bit more generic) could also be an `espejote.libsonnet` helper